website/docs/interactors/4-matchers-assertions.md

```diff
@@ -0,0 +1,267 @@
+
+```js
+TextField().exists({ placeholder: 'USERNAME' });
+```
+
+### Using `absent()` for non-existant results
+
+If you want to assert that something is not visible on the page, use `absent()`. For example, when someone is logged in, maybe the sign in button should be hidden:
+
+```js
+Button('Sign In').absent();
+```
+
+## Writing flexible assertions using Matchers
+
+Matchers are methods that let you add logic or flexibility when you are searching for elements. They can be used within locators, filters, and assertions. In this section, you will learn how to use matchers to write better tests.
+
+Matchers include the following methods:
+
+- Text flexibility like `including()` and `matching()`
+- Conditionals like `and()`, `or()`, and `not()`
+- Iterables like `some()` and `every()`
+
+### When to use Matchers
+
+If your tests are written against a simulated database, it might not be important what the randomized users' names are. Maybe you just want to assert that a login success message includes the word "welcome", such as "Welcome, Taylor!" Or perhaps it does not matter what the name is, but you still want to assert that a name is displayed. These are situations where matchers would come in handy.
+
+### including, matching
+
+The `including()` matcher invokes Javascript's `includes()` String method to check if the argument is included in the value of your interactors' locator or filter.
+
+And `matching()` is for when you want to use regular expression instead of a string.
+
+To demonstrate how you can use these two matchers, let's take this Heading element:
+
+```html
+<h1>Foo Bar</h1>
+```
+
+Here is how you would use `including()` and `matching()` to locate the header and assert that it exists:
+
+```js
+Heading(including('Foo')).exists();
+Heading(matching(/Bar$/)).exists();
+```
+
+### and, or, not
+These next three matchers: `and()`, `or()`, and `not()` are different from the first two matchers in that they can take multiple arguments and the arguments can be either a value or another matcher.
+
+For the next few examples, we'll be using these two link elements:
+
+```html
+<a href="https://google.com">Google</a>
+<a href="https://twitter.com">Twitter</a>
+```
+
+Let's first see how you can combine `and()` with some of the other matchers. In this test we are checking to see if there is a link with a `href` property that starts with `https`, includes `google`, _and_ ends with `.com`:
+
+```js
+Link({ 
+  href: and(
+    matching(/^https/),
+    including('google'),
+    matching(/\.com$/)
+  )
+}).exists();
+```
+
+If there is no link that matches all three conditions, the test would fail on account of the interactor not returning any elements.
+
+In this next example, we will pass in one false value to `or()`:
+
+```js
+Link('Google').has({ 
+  href: or(
+    'https://google.com',
+    'https://twitter.com'
```

Comment:
Just for the sake of making the example more realistic, maybe the second option could be `https://google.ca` or another domain (which makes me wonder, can `or` take an arbitrary amount of parameters?)

Reply:
We can change up the example however you'd like. Any specific reason for `https://google.ca`?

`or` can take more than two arguments:

> These next three matchers: and(), or(), and not() are different from the first two matchers in that they can take multiple arguments and the arguments can be either a value or another matcher.

Reply:
No reason, just that a link to `Google` probably has a `google` href. How about this:

```
Link('Google').has({ 
  href: or(
    'https://google.com',
    'https://google.co.uk',
    'https://google.es'
...
```

Reply:
Oh I see what you're saying. We're straying off into an unrealistic example. However, I do like that the two are very different and it makes it easier to get the message across. If anybody is skimming through the docs, they might not notice the difference between `.com` and `.ca`.

So when we take this example:
```js
Link('Google').has({
  href: or(
    'https://google.com',
    'https://twitter.com'
  )
})
```

It's very clear without any context that `or()` is a conditional that only needs one condition to return true. The reader doesn't even need to remember the HTML layout to understand the example:
```html
<a href="https://google.com">Google</a>
<a href="https://twitter.com">Twitter</a>
```

Whereas with:
```js
Link('Google').has({
  href: or(
    'https://google.com',
    'https://google.co.uk'
  )
})
```

The reader might go "why not?" It's a silly example but matchers are used in complex situations. Because interactors are already so intuitive that the need for matchers would, I assume, be in special cases. Meaning I think it's appropriate to show unrealistic examples as long as it helps the readers understand how they work in the quickest way possible.

Reply:
I like the example showing 3 TLDs. I think when there are three, it's easier to spot the difference. The other reason I like it is that I would not think to pass more than 2 parameters into this, and so it's teaching something useful about the API without introducing any new concepts/commentary.

This example is an opportunity to show off Frontside instead of google, potentially:
```
https://frontside.com/
https://www.frontside.com/
```

Reply:
@jenweber 

> The other reason I like it is that I would not think to pass more than 2 parameters into this, and so it's teaching something useful about the API without introducing any new concepts/commentary.

The first conditional matcher example we use takes 3 arguments. I did this intentionally because I also assumed people would instinctively limit themselves to 2 arguments otherwise.

> I like the example showing 3 TLDs.

Are you saying our example should have three links?

```html
<a href="https://frontside.com">Frontside</a>
<a href="https://google.com">Google</a>
<a href="https://twitter.com">Twitter</a>
```

Three links is a bit overkill and over complicates the examples without adding more value in my opinion. For `and` and `or` we can get away with using just one link:

```js
and(
  true1(),
  true2(),
  true3()
)
```

```js
Link().has({
  or(
    'frontside.com',
    'whatever random href although taking from the second link would make the example clearer'
})
```

But for `not`, it makes more sense to have two different links. I'm alternating the usage of matchers between assertions and in the interactor so people don't get stuck using them for one or the other. The `and` example uses matchers for the filter of the interactor and `or` uses matchers for the assertion. It's not really practical or realistic to use `and` or `or` for the locator in these particular examples. But in order to use `not` in the locator, we need a second link or else the assertion becomes pointless:

```js
Link(not('Frontside')).has({ href: 'whatever' }); // this would fail because we don't have a second link
Link(not('whatever')).has({ href: 'https://frontside.com' }); // this would be a pointless test
```

So we can't get away with one and three is too many.

Reply:
I don't think I follow your argument Min. I thought Jen was suggesting something like this:

```js
Link('Frontside').has({
  href: or(
    'https://frontside.com',
    'https://www.frontside.com',
    'https://frontside.io'
  )
})
```

Reply:
@jorgelainfiesta For the `or()` matcher we just need to illustrate that one false statement can still make a test pass because it's an either-or situation.

```js
Link('Frontside').has({
  href: or(
    'frontside',
    'frontside',
    'frontside'
  )
})
```

☝️ isn't as clear as:

```js
Link('Frontside').has({
  href: or(
    'frontside',
    'not-frontside'
  )
})
```

Was the point I was trying to make.

And also do we actually add the different TLDs into the HTML example?

```html
<a href="https://frontside.com">Frontside</a>
<a href="https://www.frontside.com">Frontside</a>
<a href="https://www.frontside.io">Frontside</a>
```

Because with only those three links, it's hard to discuss `not()` was the other argument I was making.

Reply:
I was thinking that the HTML example could shoot show a single link, and the “or” matcher includes three different URLs. International TLDs is a more realistic example than frontside urls, but the frontside URLs get the job done too. I think either example works here.

Reply:
I still think the non-realistic example drives the point better of how `or()` is used than using `google.ca` and `google.co.uk`.

Because if I was reading this, I'll be thinking, "why not just use `href: including('google')`? what are the benefits of using `or()`?". And in what real life scenario do we need to ensure that it's just one of the three of all the TLDs? It would make sense to use `not()` if we wanted to specify that it should not be `google.ca`.

So by doing `or('google.com', 'twitter.com')`, I think we all know how silly the example is but we get the point much faster.

But anyway, I changed it. You can see the commit [here](https://github.com/thefrontside/bigtest/pull/902/commits/2d8872c7521ceb6247b0079eca363309c9b0b4b2). I didn't switch one of the links to Frontside yet because people will already be on the frontside website when they're reading this guide, and putting our Frontside link next to Google's might give off an inaccurate impression of how we perceive ourselves.

Reply:
Ah gotcha, I see your point. Indeed `including` would've been a better choice indeed. What is the use of `or` then? I can't think of a use case but I guess there are plenty?

Reply:
The only real use case I saw was where someone wants to assert for a Button that may or may not be visible but still available in the DOM:
```js
Button({ visibility: or(true, false) }).is(whatever);
```

So we could change all the examples for `and`, `or`, and `not` so it uses a button instead but that boolean for visibility really only applies to `or` so there's a lot of switching back and forth in the examples even if we use the same Button.

Like you mentioned below, let's just brush this off for now and update it later.